clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/provider/agent/DockerRegistryImageCachingAgent.groovy

```diff
@@ -131,8 +131,10 @@ class DockerRegistryImageCachingAgent implements CachingAgent, AccountAware, Age
     Map<String, DefaultCacheDataBuilder> cachedTags = DefaultCacheDataBuilder.defaultCacheDataBuilderMap()
     Map<String, DefaultCacheDataBuilder> cachedIds = DefaultCacheDataBuilder.defaultCacheDataBuilderMap()
 
-    tagMap.forEach { repository, tags ->
-      tags.forEach { tag ->
+    tagMap.entrySet().parallelStream().forEach { entry ->
```

Comment:
I found a smaller performance improvement in my tests using the outer parallel stream as well, with a GCR registry with 125 repositories and > 5000 total tags. I have no problem removing it though if this may have side effects.

Reply:
Do you have data on the performance improvements you saw from also making the outer stream parallel?  (Ideally isolated to just calling this particular method to avoid effects of other code running at startup.)  From inspecting the code, the overhead of splitting up the `entrySet` and farming the work out to different threads will likely far outweigh the benefit of parallelizing the two assignments in that block.

There's some general guidance on when to use parallel stream [here](http://gee.cs.oswego.edu/dl/html/StreamParallelGuidance.html).  The inner use of `parallelStream` meets that guidance (largely because it is parallelizing blocking i/o operations) but the use case does not.

Reply:
Thanks for that reference!, the difference between nested parallel streams vs only the inner one don't seem to make much of a difference , I did another test with these results (only one docker registry with 125 different repositories, no cloud accounts, slow internet connection of 3MB, running on a MacBook Pro):

Nested parallel streams:
```
2019-08-12 12:28:23.145  INFO 96126 --- [ecutionAction-3] .d.r.p.a.DockerRegistryImageCachingAgent : Describing items in gcr/DockerRegistryImageCachingAgent[1/1]
2019-08-12 12:38:02.931  INFO 96126 --- [ecutionAction-3] .d.r.p.a.DockerRegistryImageCachingAgent : Caching 5127 tagged images in gcr/DockerRegistryImageCachingAgent[1/1]
2019-08-12 12:38:02.934  INFO 96126 --- [ecutionAction-3] .d.r.p.a.DockerRegistryImageCachingAgent : Caching 5127 image ids in gcr/DockerRegistryImageCachingAgent[1/1]
```
Cache processing time: 9:39 minutes

Only inner parallel stream:
```
2019-08-12 13:00:52.728  INFO 96493 --- [ecutionAction-3] .d.r.p.a.DockerRegistryImageCachingAgent : Describing items in gcr/DockerRegistryImageCachingAgent[1/1]
2019-08-12 13:11:37.635  INFO 96493 --- [ecutionAction-3] .d.r.p.a.DockerRegistryImageCachingAgent : Caching 5127 tagged images in gcr/DockerRegistryImageCachingAgent[1/1]
2019-08-12 13:11:37.637  INFO 96493 --- [ecutionAction-3] .d.r.p.a.DockerRegistryImageCachingAgent : Caching 5127 image ids in gcr/DockerRegistryImageCachingAgent[1/1]
```
Cache processing time: 10:45 minutes

In both cases, 8 `ForkJoinPool` threads were created. I think the benefits of parallelism in outer vs inner loop kind of depends if the registry has 5000 repositories with one tag, vs 1 repository with 5000 tags. As usually there are more tags than repositories it makes sense that inner loop takes most of the work.